File: Replace-Words-Trie.java
\class Solution {
    
    class TrieNode {
        boolean isWord;
        String str;
        TrieNode[] children;
        
        public TrieNode(){
            isWord = false;
            str = "";
            children = new TrieNode[26];
        }
    }
    
    TrieNode root;
    
    public String replaceWords(List<String> dict, String sentence) {
        
        root = new TrieNode();
        
        // create Trie with dic
        create_trie(dict);
        
        String[] str = sentence.split(" ");
        
        String ret_val = "";
        
        for (String w : str){
            
            ret_val += new_str(w) + " ";
        }
        
        return ret_val.substring(0,ret_val.length() - 1);
    }
    
    private String new_str(String str){
        TrieNode node = root;
        
        for (int x = 0; x < str.length(); x++){
            
            Character ch = str.charAt(x);
            
            if (node.isWord == true){
                return node.str;
            }
            
            if (node.children[ch -'a'] == null){
                break;
            } else {
                node = node.children[ch -'a'];
            }
            
        }
        
        return str;
    }
    
    private void create_trie(List<String> dict){
        
        
        for (String word : dict){
            TrieNode node = root;
            
            for (int x = 0; x < word.length(); x++){
                 Character ch = word.charAt(x);
                if (node.children[ch - 'a'] == null){
                    node.children[ch - 'a'] = new TrieNode();
                }
                node = node.children[ch - 'a'];
            }
           node.isWord = true;
           node.str = word; 
        }
    }
}

File: replaceWordsTrie.py
# Time complexity:  O(M*N) where M is length of the dict and N is number of words in the sentence
# Space complexity: O(N) where N is length of the sentence
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class TrieNode:
    def __init__(self):
        self.children = dict()
        self.isWord = False


class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for letter in word:
            child = node.children.get(letter)
            if child is None:
                child = TrieNode()
                node.children[letter] = child
            node = child
        node.isWord = True

    def search(self, word):
        answer = ''
        node = self.root
        for letter in word:
            node = node.children.get(letter)
            if node is None:
                break
            answer += letter
            if node.isWord:
                return answer
        return word


class Solution(object):
    def replaceWords(self, dict, sentence):
        """
        :type dict: List[str]
        :type sentence: str
        :rtype: str
        """
        trie = Trie()
        for word in dict:
            trie.insert(word)
        answer = []
        for word in sentence.split():
            answer.append(trie.search(word))
        return ' '.join(answer)
